dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/rest/rest/client/user/MessageAndRestClientTestServlet.java

```diff
@@ -0,0 +1,145 @@
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.microprofile.telemetry.internal_fat.apps.rest.rest.client.user;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.junit.Test;
+
+import componenttest.app.FATServlet;
+import io.openliberty.microprofile.telemetry.internal_fat.common.SpanDataMatcher;
+import io.openliberty.microprofile.telemetry.internal_fat.common.spanexporter.InMemorySpanExporter;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import javax.inject.Inject;
+import javax.servlet.annotation.WebServlet;
+import junit.framework.Assert;
+
+@WebServlet("/testMessageAndRestClientTestServlet")
+public class MessageAndRestClientTestServlet extends FATServlet {
+
+    private static final long MAX_WAIT = 40000; //millis
+    private static final long WAIT_INTERVAL = 50; //millis
+
+    private static Map<Span, Integer> spanAndExpectedNumber = new HashMap<Span, Integer>();
+
+    @Inject
+    private SimpleReactiveMessagingBean reactiveBean;
+
+    @Inject
+    InMemorySpanExporter inMemorySpanExporter;
+
+    @Test
+    public void testReactiveMessagingThatCallsJaxClient() throws InterruptedException {
+        List<String> results = getResultsWithWaitIfNeeded();
+
+        assertEquals(3, results.size());
+        assertEquals(3, results.size());
+        assertFalse(results.contains("LENGTH 8"));
+        assertFalse(results.contains("LENGTH 9!"));
+        assertTrue(results.contains("LENGTH 10!"));
+        assertTrue(results.contains("LENGTH 11!!"));
+        assertTrue(results.contains("LENGTH 12!!!"));
+    }
+
+    @Test
+    public void testSpanInBothJaxCallerAndCalled() throws InterruptedException {
+        getResultsWithWaitIfNeeded();//we just want to delay until the results are in.
+        List<List<SpanData>> listOfListOfResults = new ArrayList<List<SpanData>>(5);
+
+        SpanDataMatcher spanInsideClientMatcher = SpanDataMatcher.hasName("SimpleReactiveMessagingBean.toUpperCase");
+
+        SpanDataMatcher spanInsideJaxServiceMatcher = SpanDataMatcher.hasName("RestClientThatCapitalizes.jaxServiceThatCapitalizes");
+
+        //Each message will end up with four spans being registered. And there are five messages.
+        for (Span s : spanAndExpectedNumber.keySet()) {
+            listOfListOfResults.add(new ArrayList(inMemorySpanExporter.getFinishedSpanItems(spanAndExpectedNumber.get(s), s)));
+        }
```

Comment:
I don't think you want the code under test reporting how many spans are expected. 

I think if the code under test didn't run for some reason, the test wouldn't expect any spans and would pass without checking anything.

Reply:
Hmm, I guess you need to know the traceId to query the `inMemorySpanExporter`.

I'd suggest just collecting a list of spans (or traceIds), and check it's the expected length before calling `getFinishedSpanItems`.

I would get rid of the expected number part, it seems to always be 4...